packages/mermaid/src/diagrams/xychart/parser/xychart.jison

```diff
@@ -0,0 +1,201 @@
+  | X_AXIS parseXAxis
+  | Y_AXIS parseYAxis
+  | LINE parseLineData                    { yy.setLineData({text: '', type: 'text'}, $parseLineData); }
+  | LINE text parseLineData               { yy.setLineData($text, $parseLineData); }
+  | BAR parseBarData                      { yy.setBarData({text: '', type: 'text'}, $parseBarData); }
+  | BAR text parseBarData                 { yy.setBarData($text, $parseBarData); }
+	;
+
+parseLineData
+  : SQUARE_BRACES_START NUMBER_WITH_DECIMAL parseLineData   {$parseLineData.unshift(Number($NUMBER_WITH_DECIMAL)); $$ = $parseLineData}
+  | COMMA NUMBER_WITH_DECIMAL parseLineData                 {$parseLineData.unshift(Number($NUMBER_WITH_DECIMAL)); $$ = $parseLineData;}
+  | SQUARE_BRACES_END                                       {$$ = []}
+  ;
+
+parseBarData
+  : SQUARE_BRACES_START NUMBER_WITH_DECIMAL parseBarData    {$parseBarData.unshift(Number($NUMBER_WITH_DECIMAL)); $$ = $parseBarData}
+  | COMMA NUMBER_WITH_DECIMAL parseBarData                  {$parseBarData.unshift(Number($NUMBER_WITH_DECIMAL)); $$ = $parseBarData;}
+  | SQUARE_BRACES_END                                       {$$ = []}
+  ;
+
+parseXAxis
+  : text                                                          {yy.setXAxisTitle($text);}
+  | text xAxisBandData                                            {yy.setXAxisTitle($text); yy.setXAxisBand($xAxisBandData);}
+  | text NUMBER_WITH_DECIMAL ARROW_DELIMITER NUMBER_WITH_DECIMAL  {yy.setXAxisTitle($text); yy.setXAxisRangeData(Number($2), Number($4));}
+  ;
+
+xAxisBandData
+  : SQUARE_BRACES_START text xAxisBandData                  {$xAxisBandData.unshift($text); $$ = $xAxisBandData}
+  | COMMA text xAxisBandData                                {$xAxisBandData.unshift($text); $$ = $xAxisBandData;}
+  | SQUARE_BRACES_END                                       {$$ = []}
+  ;
+
+parseYAxis
+  : text  {yy.setYAxisTitle($text);}
+  | text NUMBER_WITH_DECIMAL ARROW_DELIMITER NUMBER_WITH_DECIMAL  {yy.setYAxisTitle($text); yy.setYAxisRangeData(Number($2), Number($4));}
+  ;
+
+directive
+  : openDirective typeDirective closeDirective
+  | openDirective typeDirective ':' argDirective closeDirective
+  ;
+
+eol
+  : NEWLINE
+  | SEMI
+  | EOF
+  ;
+
+openDirective
+  : open_directive { yy.parseDirective('%%{', 'open_directive'); }
+  ;
+
+typeDirective
+  : type_directive { yy.parseDirective($1, 'type_directive'); }
+  ;
+
+argDirective
+  : arg_directive { $1 = $1.trim().replace(/'/g, '"'); yy.parseDirective($1, 'arg_directive'); }
+  ;
+
+closeDirective
+  : close_directive { yy.parseDirective('}%%', 'close_directive', 'xychart'); }
+  ;
+
+text: alphaNum
+    { $$={text:$alphaNum, type: 'text'};}
+    | STR
+    { $$={text: $STR, type: 'text'};}
+    | MD_STR
+    { $$={text: $MD_STR, type: 'markdown'};}
+    ;
+
+alphaNum
+    : alphaNumToken
+    {$$=$alphaNumToken;}
+    | alphaNum alphaNumToken
+    {$$=$alphaNum+''+$alphaNumToken;}
+    ;
```

Comment:
I remember @ibrahimWassouf [was dealing with those grammar rules](https://github.com/mermaid-js/mermaid/pull/4657/files#diff-0d402731a094263cd9703f40974d74949df8b287c03dc68f47c317a6ec8a35dcR583-R597), please check this part against his recent improvements

@ibrahimWassouf it would be nice to get a review from you on this grammar

Reply:
@nirname @subhash-halder From what I had to go through to keep backwards compatibility while fixing the bugs, I would argue against pushing text states without  demarcating them with something (similar to how string are demarcated with quotation marks). For now, what's in this PR works because the `title` token has the highest precedence. However, say that we wanted to add inline styling or anything after the title text, without giving it too much precedence it would be impossible. E.g., imagine we were going to implement a classDiagram inline styling in the future. It would be impossible to do something like  `title some text:::styleclass` without giving the `:::` token higher precedence, because the alphanum/text token can go on until it hits a newline. This would not be a problem if we only used strings or markdown strings for title text, because as soon as we read the end qoutation mark, we can do anything afterwards

Reply:
@nirname @ibrahimWassouf I have change the implementation of the title, please let me know if it's ok now or I need to do any changes.